File: frontend/lib/sensor.dart
class Sensor {
  const Sensor({
    required this.id,
    required this.name,
    required this.unit,
    required this.health,
    this.latestReading,
  });

  factory Sensor.fromJson(Map<String, dynamic> json) {
    final latestReading = json['latest_reading'];
    return Sensor(
      id: json['id'] as String,
      name: json['name'] as String,
      unit: json['unit'] as String,
      health: json['health'] as String,
      latestReading: latestReading == null
          ? null
          : LatestReading.fromJson(latestReading as Map<String, dynamic>),
    );
  }

  final String id;
  final String name;
  final String unit;
  final String health;
  final LatestReading? latestReading;
}

class LatestReading {
  const LatestReading({
    required this.recordedAt,
    required this.value,
    required this.status,
  });

  factory LatestReading.fromJson(Map<String, dynamic> json) {
    return LatestReading(
      recordedAt: DateTime.parse(json['recorded_at'] as String),
      value: (json['value'] as num).toDouble(),
      status: json['status'] as String,
    );
  }

  final DateTime recordedAt;
  final double value;
  final String status;
}

File: backend/internal/telemetry/sensor.go
package telemetry

import "time"

type ReadingStatus string

const (
	ReadingStatusValid      ReadingStatus = "valid"
	ReadingStatusOutOfRange ReadingStatus = "out_of_range"
)

type Reading struct {
	RecordedAt time.Time
	Value      float64
	Status     ReadingStatus
}

type WriteOutcome string

const (
	WriteOutcomeStored    WriteOutcome = "stored"
	WriteOutcomeDuplicate WriteOutcome = "duplicate"
	WriteOutcomeConflict  WriteOutcome = "conflict"
)

type WriteResult struct {
	Outcome        WriteOutcome
	ExistingValue  float64
	ExistingStatus ReadingStatus
}

type Sensor struct {
	ID            string
	Name          string
	Unit          string
	ValidMin      float64
	ValidMax      float64
	LatestReading *Reading
}

type Health string

const (
	HealthOK            Health = "ok"
	HealthStale         Health = "stale"
	HealthNeverReported Health = "never_reported"
)

const healthyWindow = 15 * time.Minute

func SensorHealth(latest *Reading, now time.Time) Health {
	if latest == nil {
		return HealthNeverReported
	}
	if !latest.RecordedAt.Before(now.Add(-healthyWindow)) {
		return HealthOK
	}
	return HealthStale
}

func ClassifyReading(sensor Sensor, value float64) ReadingStatus {
	if value < sensor.ValidMin || value > sensor.ValidMax {
		return ReadingStatusOutOfRange
	}
	return ReadingStatusValid
}
